master/buildbot/reporters/telegram.py

```diff
@@ -0,0 +1,1045 @@
+            ans = yield res.json()
+            if not ans.get('ok'):
+                raise ValueError("[{}] {}".format(res.code, ans.get('description')))
+        except AssertionError as err:
+            # just for tests
+            raise err
+        except Exception as err:
+            self.log("ERROR: cannot send Telegram request {}: {}".format(path, err))
+        else:
+            return ans.get('result', True)
+
+    @defer.inlineCallbacks
+    def _post_until_success(self, path, **kwargs):
+        logme = True
+        while True:
+            try:
+                res = yield self.http_client.post(path, **kwargs)
+                ans = yield res.json()
+                if not ans.get('ok'):
+                    self.log("ERROR: cannot send Telegram request {}: "
+                             "[{}] {}".format(path, res.code, ans.get('description')))
+                    return
+            except AssertionError as err:
+                # just for tests
+                raise err
+            except Exception as err:
+                msg = "ERROR: cannot send Telegram request {} (will try again): {}".format(path, err)
+                if logme:
+                    self.log(msg)
+                    logme = False
+                yield task.deferLater(self.reactor, self.retry_delay, lambda: None)
+            else:
+                return ans.get('result', True)
+
+    @defer.inlineCallbacks
+    def set_nickname(self):
+        res = yield self._post_until_success('/getMe')
+        if res:
+            self.nickname = res.get('username')
+
+    @defer.inlineCallbacks
+    def answer_query(self, query_id, notify=None):
+        params = dict(callback_query_id=query_id)
+        if notify is not None:
+            params.update(dict(text=notify))
+        return (yield self._post('/answerCallbackQuery', json=params))
+
+    @defer.inlineCallbacks
+    def send_message(self, chat, message, parse_mode='Markdown',
+                     reply_to_message_id=None, reply_markup=None,
+                     **kwargs):
+        result = None
+
+        message = message.strip()
+        while message:
+            params = dict(chat_id=chat)
+            if parse_mode is not None:
+                params['parse_mode'] = parse_mode
+            if reply_to_message_id is not None:
+                params['reply_to_message_id'] = reply_to_message_id
+                reply_to_message_id = None  # we only mark first message as a reply
+
+            if len(message) <= 4096:
+                params['text'], message = message, None
+            else:
+                n = message[:4096].rfind('\n')
+                n = n + 1 if n != -1 else 4096
+                params['text'], message = message[:n].rstrip(), message[n:].lstrip()
+
+            if not message and reply_markup is not None:
+                params['reply_markup'] = reply_markup
+
+            params.update(kwargs)
+
+            result = yield self._post('/sendMessage', json=params)
```

Comment:
why not until success for that one?

Reply:
Actually there is no reason `_post` and `_post_until_success` differ...